zerver/tests/test_send_email.py

```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
-        backend.connection.noop.return_value = [250]
-        initialize_connection(backend)
-        self.assertEqual(backend.open.call_count, 1)
-        self.assertEqual(backend.connection.noop.call_count, 1)
-
-        # Test the old connection case when it was closed by the server
-        backend.connection.noop.return_value = [404]
-        backend.close.return_value = False
-        initialize_connection(backend)
-        # 2 more calls to open, 1 more call to noop and 1 call to close
-        self.assertEqual(backend.open.call_count, 3)
-        self.assertEqual(backend.connection.noop.call_count, 2)
-        self.assertEqual(backend.close.call_count, 1)
-
-        # Test backoff procedure
-        backend.open.side_effect = OSError
-        with self.assertRaises(OSError):
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
+        self.assertEqual(backend.opened_at, timezone_now())
+
+        with (
+            mock.patch.object(backend, "_open", wraps=backend._open) as _open_call,
+            mock.patch.object(backend, "close", wraps=backend.close) as close_call,
+        ):
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
+            initialize_connection(backend)
+
+            self.assertEqual(_open_call.call_count, 1)
+            self.assertEqual(backend.connection.noop.call_count, 1)
+            close_call.reset_mock()
+            _open_call.reset_mock()
+
+            # Test the old connection case when it was closed by the server
+            backend.connection.noop.return_value = [404]
+            initialize_connection(backend)
+
+            # 2 calls to open and 1 call to close
+            self.assertEqual(_open_call.call_count, 2)
+            self.assertEqual(close_call.call_count, 1)
+            close_call.reset_mock()
+            _open_call.reset_mock()
+
+            # Test backoff procedure
+            _open_call.side_effect = OSError
+            with self.assertRaises(OSError):
+                initialize_connection(backend)
+            # 3 calls to open as we try 3 times before giving up
+            self.assertEqual(_open_call.call_count, 3)
+
+    @time_machine.travel(timezone_now(), tick=False)
+    @override_settings(EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=0)
+    def test_max_connection_lifetime_zero(self) -> None:
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
+        self.assertEqual(backend.opened_at, timezone_now())
+
+        with (
+            mock.patch.object(backend, "_open", wraps=backend._open) as _open_call,
+            mock.patch.object(backend, "close", wraps=backend.close) as close_call,
+        ):
+            # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+            initialize_connection(backend)
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
```

Comment:
I don't get this combination, 1 close and 0 noops is fine, but in case of an existing connection being open, first `_open` call will be used to check the existence of the connection, then we'll close the open connection and reopen the connection using `_open` leading to two _open calls

Reply:
Same reason for noop count as the other comment. For _open, it will be called once to check whether the connection exists or not at the beginning of the `open` function and for a second time after calling `.close()`.